### PR TITLE
Refactorings: bug & linter fixes, code simplifications

### PR DESCRIPTION
**Pull request by SchumacherFM**
- use NewReader instead of NewBuffer
- return errors directly without err != nil checks
- consistent naming of method receivers
- pre-allocate buffers and other slices
- use t.Fatal instead of panic
- use buf.ReadFrom instead of io.Copy
- for/select to for/range on channel
- rename internal functions for better understanding
- use bufio.NewReaderSize when reading from disk because AWS/GCP can
  be slow.

@loeffert please critically review, the 2nd PR depends at bit on this PR

### CODE REVIEW

File: client/sockettransport.go [hunk 1]
```diff
@@ -79,7 +79,7 @@ func (c *socketTransport) call(handler server.Handler, request interface{}, resp
 
 	// read response
 	var (
-		responseBytes  = []byte{}
+		responseBytes  = make([]byte, 0, 8192)
```
Comment (dreadl0ck) [inline client/sockettransport.go]: Whats the reason to setting the maximum buffer size to 8192 bytes?

Please implement a unit test to show that this is robust in all edge cases.

File: client/httptransport.go [hunk 2]
```diff
@@ -33,7 +33,7 @@ func (ht *httpTransport) call(handler server.Handler, request interface{}, respo
 	req, errNewRequest := http.NewRequest(
 		http.MethodPost,
 		ht.endpoint+"/"+string(handler),
-		bytes.NewBuffer(requestBytes),
+		bytes.NewReader(requestBytes),
```
Comment (dreadl0ck) [inline client/httptransport.go]: Whats the advantage of using a Reader compared to the bytes.Buffer here?

File: repo/history.go [hunk 3]
```diff
@@ -132,6 +131,6 @@ func (h *history) getCurrent(buf *bytes.Buffer) (err error) {
 		return err
 	}
 	defer f.Close()
-	_, err = io.Copy(buf, f)
+	_, err = buf.ReadFrom(f)
```
Comment (dreadl0ck) [inline repo/history.go]: ReadFrom reads data until EOF and appends it to the buffer.

We used io.Copy here because it copies the data in small chunks to keep the memory usage low.

File: repo/repo.go [hunk 4]
```diff
@@ -223,15 +224,16 @@ func (repo *Repo) GetRepo() map[string]*content.RepoNode {
...
+		return
 	}
+	defer f.Close()
 
-	w.Write([]byte("{\"reply\":"))
-	_, err = io.Copy(w, f)
-	if err != nil {
+	w.Write([]byte(`{"reply":`))
+	// bufio.NewReaderSize speeds up reading in an AWS/GCP environment
+	if _, err := io.Copy(w, bufio.NewReaderSize(f, 2*4096)); err != nil {
```
Comment (dreadl0ck) [inline repo/repo.go]: -> bufio.NewReaderSize speeds up reading in an AWS/GCP environment

Do you have benchmarks for this?

**Comment by dreadl0ck** [comment]
$ make test
go test -v ./...
# github.com/SchuhmacherFM/contentserver/server
server/handlerequest.go:58:11: undefined: responses.NewErrorf
server/handlerequest.go:65:11: undefined: responses.NewErrorf
server/handlerequest.go:68:11: undefined: responses.NewErrorf
server/socketserver.go:114:44: undefined: responses.NewErrorf

Tests for this branch fail.

File: client/sockettransport.go (hunk 1 above)
Comment (SchumacherFM) [reply client/sockettransport.go]: The 8192 is just a wild guess how large the incoming data can be. Usually you should measure it and then you can set the pre-allocated buffer. I will try to add a test.

File: client/httptransport.go (hunk 2 above)
Comment (SchumacherFM) [reply client/httptransport.go]: Less internal overhead

File: repo/history.go (hunk 3 above)
Comment (SchumacherFM) [reply repo/history.go]: Yes and not but you still read everything into `buf`, so memory would be the same. You could even grow the buffer to the size of the file ... so io.Copy has NO benefit here.

File: repo/repo.go (hunk 4 above)
Comment (SchumacherFM) [reply repo/repo.go]: I don;t have access to AWS/GCP but this is what the other bestbytes dev told me.
Also read here: https://www.datadoghq.com/blog/aws-ebs-provisioned-iops-getting-optimal-performance/

**Comment by SchumacherFM** [comment]
That is fixed in the next commit ... sorry for my fat fingers.

File: client/sockettransport.go (hunk 1 above)
Comment (SchumacherFM) [reply client/sockettransport.go]: This function is untestable in a unit test. requires to much refactoring. But you should guess yourself that pre-allocating a byte slice with a appropriate capacity avoids rebuilding the internal slice when calling `append`.

**Comment by dreadl0ck** [comment]
Rejected - please refer to my comment on buffer preallocation in the socket client.

As described in the comment, preallocating the buffer size will lead to a panic in the socket client if the size of data exceeds the buffers 8KB capacity. The size of fetched content cannot be predetermined.

For the future, please add proper unit testing for critical changes like this, and test them sufficiently prior to submitting a pull request!

Additionally, I recommend to separate 'cosmetical' code changes like integration of linter feedback / code simplifications, and changes that are targeted towards performance optimisation, into separate pull requests.

The title "Refactorings: bug & linter fixes, code simplifications" is misleading here, it hides the fact that you implemented performance tweaks that broke a fundamental part of the application.

Besides that, to speed up review of performance optimisations (such as the AWS/GCP reader tweak) and improve the codebase, make sure to add a comment with a justification and reference source for the implemented optimisation.

**Comment by SchumacherFM** [comment]
@dreadl0ck Kudos for this good critical comment! Highly appreciated! 
I will refactor it and send new PRs

**Comment by SchumacherFM** [comment]
> As described in the comment, preallocating the buffer size will lead to a panic in the socket client if the size of data exceeds the buffers 8KB capacity. The size of fetched content cannot be predetermined.

You haven't described it in the comment. Also nowhere (in a commit message nor in the source code) is it mentioned that the socket panics if the buffer size starts with 8kb pre-allocated.

> Whats the reason for setting the maximum buffer size to 8192 bytes?

also it is not a maximum buffer size. please read the Go documentation. It is the initial capacity to start with. Also `append` (https://github.com/foomo/contentserver/blob/master/client/sockettransport.go#L95) might/will reslice the `[]byte` slice to a different capacity.
